api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirEncounterServiceImpl.java

```diff
@@ -73,6 +70,57 @@ public Encounter get(@Nonnull String uuid) {
 		return result;
 	}
 	
+	@Override
+	public Encounter create(@Nonnull Encounter encounter) {
+		
+		if (encounter.getTypeFirstRep().getCodingFirstRep().getSystem() == null) {
+			throw new InvalidRequestException("Id type must be specified to update");
+		}
+		if (encounter.getTypeFirstRep().getCodingFirstRep().getSystem().equals(FhirConstants.ENCOUNTER_TYPE_SYSTEM_URI)) {
```

Comment:
So, rather than relying on the *first* type coding (especially since we want to ignore other coding types), I'd refactor this into a utility method, something like:

```java
private enum OpenmrsEncounterType {
	ENCOUNTER,
	VISIT,
	AMBIGUOUS
}

// ...

private Optional<OpenmrsEncounterType> getOpenmrsEncounterType(Encounter encounter) {
	List<OpenmrsEncounterType> openmrsEncounterTypes = encounter.getType().stream().flatMap(it -> 
		it.getCoding().stream()
			.filter(coding -> FhirConstants.ENCOUNTER_TYPE_SYSTEM_URI.equals(coding.getSystem()) ||
				FhirConstants.VISIT_TYPE_SYSTEM_URI.equals(coding.getSystem()))
			.map(coding -> {
				if (FhirConstants.ENCOUNTER_TYPE_SYSTEM_URI.equals(coding.getSystem())) {
					return OpenmrsEncounterType.ENCOUNTER;
				} else {
					return OpenmrsEncounterType.VISIT;
				}
			})
	).distinct().collect(Collectors.toList());
	
	if (openmrsEncounterTypes.isEmpty()) {
		return Optional.empty();
	} else if (openmrsEncounterTypes.size() == 1) {
		return Optional.of(openmrsEncounterTypes.get(0));
	} else {
		return Optional.of(OpenmrsEncounterType.AMBIGUOUS);
	}
}
```

Now as we end up with three states:

1. We known the encounter type of the encounter and can process it
2. No encounter type was supplied and we can throw an exception with a message to that effect
3. An encounter invalidly specified that it is *both* an OpenMRS Encounter and an OpenMRS visit and we can send an appropriate error message for that.

Plus, we don't depend on the order of the types or do anything else in response to codings we don't care about.

Reply:
@ibacher I will change the code accordingly.